rholang/src/main/scala/coop/rchain/rholang/interpreter/storage/StoragePrinter.scala

```diff
@@ -40,39 +45,79 @@ object StoragePrinter {
         }
       }.toList
     } yield {
-      if (pars.isEmpty) {
+      if (pars.isEmpty)
         "The space is empty. Note that top level terms that are not sends or receives are discarded."
-      } else {
-        val par = pars.reduce { (p1: Par, p2: Par) =>
-          p1 ++ p2
-        }
-        PrettyPrinter().buildString(par)
-      }
+      else
+        PrettyPrinter().buildString(pars.reduce(_ ++ _))
     }
 
   def prettyPrintUnmatchedSends[F[_]: FlatMap](space: RhoISpace[F]): F[String] =
+    space.toMap.map(getUnmatchedSends).map { unmatchedSends =>
+      if (unmatchedSends.isEmpty)
+        noUnmatchedSends
+      else
+        PrettyPrinter().buildString(unmatchedSends.reduce(_ ++ _))
+    }
+
+  def prettyPrintUnmatchedSends[F[_]: Concurrent](
+      deploy: DeployData,
+      runtime: Runtime[F]
+  ): F[String] = {
+    def unmatchedSends: F[List[Par]] = runtime.space.toMap.map(getUnmatchedSends)
     for {
-      mapped <- space.toMap
-      pars = mapped.map {
-        case (
-            channels: Seq[Par],
-            row: Row[BindPattern, ListParWithRandom, TaggedContinuation]
-            ) =>
-          row match {
-            case Row(data: Seq[Datum[ListParWithRandom]], _) =>
-              toSends(data)(channels)
-          }
-      }.toList
-    } yield {
-      if (pars.isEmpty) {
-        "No unmatched sends."
-      } else {
-        val par = pars.reduce { (p1: Par, p2: Par) =>
-          p1 ++ p2
-        }
-        PrettyPrinter().buildString(par)
+      checkpoint <- runtime.space.createCheckpoint()
+      beforeEval <- unmatchedSends
+      _ <- {
+        implicit val c = runtime.cost
+        Interpreter[F].evaluate(runtime, deploy.term, NormalizerEnv(deploy))
       }
+      afterEval <- unmatchedSends
+      diff      = afterEval.diff(beforeEval)
+      _         <- runtime.space.reset(checkpoint.root)
+    } yield {
+      if (diff.isEmpty)
+        noUnmatchedSends
+      else
+        PrettyPrinter().buildString(diff.reduce(_ ++ _))
     }
+  }
+
+  def prettyPrintUnmatchedSends[F[_]: Concurrent](
+      deploys: Seq[DeployData],
+      runtime: Runtime[F]
+  ): F[String] =
+    deploys
+      .map(
+        deploy =>
+          prettyPrintUnmatchedSends(deploy, runtime)
+            .map(unmatchedSends => (deploy, unmatchedSends))
+      )
+      .toList
```

Comment:
I assume if you need to traverse the collection so you can use toStream instead of toList. That would be better because it doesn't build the intermediary list